### PR TITLE
[FIX] mail: notification access rights

### PR DESCRIPTION
**Pull request by nda-odoo**
Check if user has access to the record before displaying a notification about it.

_steps to reproduce:_
- create two companies (company_A and company_B)
- using the company switcher, select only company_A
- create a partner and set the field company_id on company_A
- send a sms or a mail with an error on this partner (you should have a notification about the error on the top bar)
- use the company switcher to select only company_B

(on runbot you can simply set a company_id on one of the contact with a sms error)

_before this commit:_
an error is raised when failed messages are fetched
`odoo.exceptions.AccessError: Due to security restrictions, you are not allowed to access 'Contact' (res.partner) records.
`

_after this commit:_
only notifications about records that can be accessed by the user are displayed

opw-3558373

---
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

File: addons/mail/models/mail_message.py [hunk 1]
```diff
@@ -1142,9 +1143,7 @@ def _message_notification_format(self):
             'notifications': message.notification_ids._filtered_for_web_client()._notification_format(),
         } for message in self]
 
-    def _notify_message_notification_update(self):
-        """Send bus notifications to update status of notifications in the web
-        client. Purpose is to send the updated status per author."""
+    def _filter_can_read_record(self):
```
Comment (thiryr) [inline addons/mail/models/mail_message.py:1146]: Should be in alphabetical order, ideally

File: addons/mail/models/mail_message.py [hunk 2]
```diff
@@ -1142,9 +1143,7 @@ def _message_notification_format(self):
...
         messages = self.env['mail.message']
         for message in self:
             # Check if user has access to the record before displaying a notification about it.
```
Comment (thiryr) [inline addons/mail/models/mail_message.py:1149]: Comment seems too specific now. I'd add a docstring to the method and remove references to notifications in the docstring.

And let's keep the FIXME here I guess

**Review by reth-odoo — COMMENTED** [review]
I'm mildly worried about performance but I don't think there are many other options.. I think most of the checks are cached so it should be ok.

File: addons/mail/models/mail_message.py (hunk 1 above)
Comment (nda-odoo) [reply addons/mail/models/mail_message.py:1146]: I'm not sure of what you mean. What should be in alphabetical order? The method `_filter_can_read_record` in `addons/mail/models/mail_message.py`? Or the filtered messages?

**Comment by nda-odoo** [comment]
> I'm mildly worried about performance but I don't think there are many other options.. I think most of the checks are cached so it should be ok.

Thanks for the review!

I agree that this loop does not look good. But since it's in a separate call, it will not slow down the whole UI. The "error thingy" might just be a little slower. And since it's only maximum 100 records, it should be ok. And if doing access checks on 100 records gets really slow, what that probably mean is that another model is getting slow, not really `mail.message`.

**Review by reth-odoo — COMMENTED** [review]
Seems I never submitted this

**Comment by nda-odoo** [comment]
Ok! I changed the position of the method.

**Review by reth-odoo — APPROVED** [review]
lgtm thanks :)

@tde-banana-odoo :wave: hi, it's ping season

**Comment by reth-odoo** [comment]
For tracking
Closed as 14.0 is no longer supported and the issue is seemingly already fixed in 15.0 and up

Thanks for the work anyway!

**Pull request closed without merging**